ACPs/118-warp-signature-request/README.md

```diff
@@ -0,0 +1,81 @@
+Author(s): Cam Schultz (https://github.com/cam-schultz)
+Discussions-To: 
+Status: Proposed
+Track: Standards Track
+```
+
+## Abstract
+
+Proposes handling Warp signature requests from peers at the protocol level, rather than leaving it to VMs to implement. This will standardize the signature request interface, allowing for more flexible, VM-agnostic signature aggregation implementations.
+
+## Motivation
+
+Warp message signatures consist of an aggregate BLS signature composed of the individual signatures of a subnet's validators. Individual signatures need to be retreivable by the party that wishes to construct an aggregate signature. At present, this is left to VMs to implement, as is the case with [Subnet EVM](https://github.com/ava-labs/subnet-evm/blob/v0.6.6/plugin/evm/message/signature_request.go#20) and [Coreth](https://github.com/ava-labs/coreth/blob/v0.13.6-rc.0/plugin/evm/message/signature_request.go#L20)
+
+This creates friction in applications that are intended to operate across many VMs (or distinct implementations of the same VM). As an example, the reference Warp message relayer implementation, [awm-relayer](https://github.com/ava-labs/awm-relayer), fetches individual signatures from validators and aggregates them before sending the Warp message to its destination chain for verification. However, Subnet EVM and Coreth have distinct codecs, requiring the relayer to [switch](https://github.com/ava-labs/awm-relayer/blob/v1.4.0-rc.0/relayer/application_relayer.go#L372) according to the target codebase.
+
+Another example is ACP-75, which aims to implement acceptance proofs using Warp. The signature aggregation mechanism is not [specified](https://github.com/avalanche-foundation/ACPs/blob/main/ACPs/75-acceptance-proofs/README.md#signature-aggregation), which is a blocker for that ACP to be marked implementable.
+
+Standardizing the Warp Signature Request format by implementing it as a protocol-level (AvalancheGo) AppRequest type would simplify the implementation of ACP-75, and streamline signature aggregation for out-of-protocol services such as Warp message relayers.
+
+## Specification
+
+Subnet EVM and Coreth currently handle Warp message signature requests from peers by decoding the `AppRequest` bytes (forwarded from Avalanchego via the `AppHandler` interface) using a type registered with the message codec:
+
+```
+type MessageSignatureRequest struct {
+	MessageID ids.ID `serialize:"true"`
+}
+```
+
+`MessageID` represents the VM-agnostic Warp message ID, and could be ported as-is to the protocol. To do so, we would [define](https://github.com/ava-labs/avalanchego/blob/v1.11.10-status-removal/proto/p2p/p2p.proto) a new P2P `Message` type:
+
+```diff
+message Message {
+    ...
+    // App messages:
+    AppRequest app_request = 30;
+    AppResponse app_response = 31;
+    AppGossip app_gossip = 32;
+    AppError app_error = 34;
+    
++   // Warp signature messages:
++   MessageSignatureRequest = 35;
```

Comment:
What would the downside be of expanding this to `SignatureRequest` and having it defined as below?

```
message SignatureRequest {
    uint32 type_id = 1
    bytes id = 2
}
```

Perhaps a better name for the second `id` field should be decided but this would then generalize to both message and hash signature requests. 

You mention below in `Open Questions` that interpreting the ID could be left to the VM but in that case we would have  a naming clash where current `AppRequest` based implementation distinguishes based on the message type and has both `MessageSignatureRequest` as well as `BlockSignatureRequest`.

[Warp docs](https://github.com/ava-labs/avalanchego/tree/576b3927c54b0e7ad91602572f4aa8f9fe24d768/vms/platformvm/warp/payload) specify that `Hash` payloads are chain specific, but we wouldn't lose this generality in the proto definition. I.e. if another chain used the `Hash` `typeID` of `0x 0x00000000` for something other than the block hash this message type would still be usable for it and would disambiguate from `AddressedCall` messages.

Reply:
Maybe `signable_id`?

Reply:
I added a new `HashSignatureRequest` type that encodes a hash over arbitrary bytes. That would be a candidate for extending with a `type_id` field that's interpretable by VMs. I think `MessageSignatureRequest` should be tightly scoped to Warp message IDs specifically.

Reply:
Alternatively, we could generalize `HashSignatureRequest` to `SignatureRequest` that specifies arbitrary bytes, and leave it to the VM to define how to decode. Not sure how useful that would be over decoding the `AppRequest` payload directly.

Reply:
I agree now that we should be tightly scoped to Warp Message IDs specifically. Now that I think about it again, my initial reservations were influenced by the current `subnet-evm` implementation and I was parsing `MessageSignatureRequest` and `BlockSignatureRequest` as disjoint use-cases but really `MessageSignatureRequest` is the general one and `BlockSignatureRequest` is just a special case. 

Given that I'm fine with completely dropping my proposal of `type_id` fields and I'm not even convinced that we need the `HashSignatureRequest` either.